sycl/source/detail/pi.cpp

```diff
@@ -37,22 +38,55 @@ std::string platformInfoToString(pi_platform_info info) {
 // Check for manually selected BE at run-time.
 bool useBackend(Backend TheBackend) {
   static const char *GetEnv = std::getenv("SYCL_BE");
-  static const Backend Use =
-    std::map<std::string, Backend>{
-      { "PI_OPENCL", SYCL_BE_PI_OPENCL },
-      { "PI_OTHER",  SYCL_BE_PI_OTHER }
+  static const Backend Use = std::map<std::string, Backend>{
+      {"PI_OPENCL", SYCL_BE_PI_OPENCL}, {"PI_OTHER", SYCL_BE_PI_OTHER}
       // Any other value would yield PI_OPENCL (current default)
-    }[ GetEnv ? GetEnv : "PI_OPENCL"];
+  }[GetEnv ? GetEnv : "PI_OPENCL"];
   return TheBackend == Use;
 }
 
 // Definitions of the PI dispatch entries, they will be initialized
 // at their first use with piInitialize.
-#define _PI_API(api) decltype(::api) * api = nullptr;
+#define _PI_API(api) decltype(::api) *api = nullptr;
 #include <CL/sycl/detail/pi.def>
 
-// TODO: implement real plugins (ICD-like?)
-// For now this has the effect of redirecting to built-in PI OpenCL plugin.
+// Find the plugin at the appropriate location and return the location in
+// PluginPath
+// TODO: Change the function appropriately when there are multiple plugins.
+bool findPlugin(std::string &PluginPath) {
+  // TODO: Based on final design discussions, change the location where the
+  // plugin must be searched; how to identify the plugins etc. Currently the
+  // search is done for libpi_opencl.so file in LD_LIBRARY_PATH env only.
+  PluginPath = PLUGIN_NAME;
+  return true;
+}
+
+// Load the Plugin by calling the OS dependent library loading call.
+// Return the handle to the Library.
+void *loadPlugin(std::string PluginPath) { return loadOsLibrary(PluginPath.c_str()); }
+
+// Binds all the PI Interface APIs to Plugin Library Function Addresses.
+// TODO: Remove the 'OclPtr' extension to PI_API.
+bool bindPlugin(void *Library) {
```

Comment:
add TODO that this will eventually be a 1-to-many mapping, and that plugins bindings should set up needed infrastructure for routine PI calls into appropriate plugins.

Reply:
Probably part of a separate PR on this topic, but the 1-to-many mapping will require switching the dispatch table depending on the backend associated with a given object. However, the rest of the runtime is not aware of multiple backends, it simply calls PI_CALL functions. Does the routing to the right backend happens inside the PI API , or is it done from the SYCL runtime?

Reply:
> Probably part of a separate PR on this topic, but the 1-to-many mapping will require switching the dispatch table depending on the backend associated with a given object. However, the rest of the runtime is not aware of multiple backends, it simply calls PI_CALL functions. Does the routing to the right backend happens inside the PI API , or is it done from the SYCL runtime?

Yes. This change will be done in the patches that follow. I think we will do it in the SYCL runtime, based on the context.

Reply:
> add TODO that this will eventually be a 1-to-many mapping, and that plugins bindings should set up needed infrastructure for routine PI calls into appropriate plugins.

done.